src/base/static/components/form-field/index.js

```diff
@@ -1,354 +1,114 @@
+    const autofillValue = this.props.fieldConfig.autocomplete
+      ? Util.getAutocompleteValue(this.props.fieldConfig.name)
       : null;
-
-    fieldConfig.hasAutofill = !!autofillValue;
-
-    let fieldValue;
-    switch (fieldConfig.type) {
-      case constants.BIG_TOGGLE_FIELD_TYPENAME:
-        fieldValue =
-          autofillValue ||
-          fieldConfig.default_value ||
-          fieldConfig.content[1].value; // "off" position of the toggle
-        break;
-      case constants.BIG_CHECKBOX_FIELD_TYPENAME:
-        fieldValue =
-          fromJS(autofillValue) ||
-          fromJS(fieldConfig.default_value) ||
-          ImmutableList();
-        break;
-      case constants.PUBLISH_CONTROL_TOOLBAR_TYPENAME:
-        fieldValue = autofillValue || fieldConfig.default_value;
-        break;
-      case constants.MAP_DRAWING_TOOLBAR_TYPENAME:
-        fieldValue = null;
-        break;
-      default:
-        fieldValue = autofillValue || fieldConfig.default_value || "";
-        break;
-    }
-
-    const fieldObj = ImmutableMap()
+    this.props.fieldConfig.hasAutofill = !!autofillValue;
+    const initialFieldValue = this.fieldDefinition.getInitialValue(
+      autofillValue,
+      this.props.fieldConfig.default_value,
+      this.props.fieldConfig
+    );
+    const initialFieldState = ImmutableMap()
       .set(constants.FIELD_STATE_INITIALIZED_KEY, true)
-      .set(constants.FIELD_STATE_VALUE_KEY, fieldValue)
-      .set(constants.FIELD_STATE_FIELD_TYPE_KEY, fieldConfig.type)
+      .set(constants.FIELD_STATE_VALUE_KEY, initialFieldValue)
+      .set(constants.FIELD_STATE_FIELD_TYPE_KEY, this.props.fieldConfig.type)
+      .set(
+        constants.FIELD_RENDER_KEY,
+        this.props.fieldState.get(constants.FIELD_RENDER_KEY)
+      )
       .set(
         constants.FIELD_STATE_VALIDITY_KEY,
-        this.validator.validate(
-          fieldValue,
+        this.fieldDefinition.validator.validate(
+          initialFieldValue,
           this.props.places,
           this.props.landmarks
         )
       )
-      .set(constants.FIELD_STATE_VALIDITY_MESSAGE_KEY, this.validator.message);
+      .set(
+        constants.FIELD_STATE_VALIDITY_MESSAGE_KEY,
+        this.fieldDefinition.validator.message
+      );
 
-    return fieldObj;
+    this.props.onFieldInitialize(
+      this.props.fieldConfig.name,
+      initialFieldState
+    );
   }
 
-  buildField(fieldConfig) {
-    // TODO: appropriate field visibility based on admin status
-
-    const cn = {
-      optionalMsg: classNames("input-form__optional-msg", {
-        "input-form__optional-msg--visible": fieldConfig.optional,
-      }),
-    };
-    const fieldPrompt = (
-      <p className="input-form__field-prompt" key={1}>
-        {fieldConfig.prompt}
-        <span className={cn.optionalMsg}>{messages.optionalMsg}</span>
-      </p>
+  shouldComponentUpdate(nextProps) {
+    return (
+      nextProps.isInitializing ||
```

Comment:
Why do we need this? Does this component not re-render when `this.props.selectedCategoryConfig.fields` updates?